### DISCUSSION TITLE
Association terms in PC-SAFT

### DISCUSSION THREAD

**Discussion by cassguff** [usnistgov/teqp#157] [General]
I know you haven't included the association terms for PC-SAFT in teqp, and as was stated in the teqp paper this was because it would be necessary to iterate over these terms. Just to satisfy my curiosity, can anyone expand on this for me? Why are iterations so dangerous in taking derivatives of equations of state? Is it just complicated to code, or does it create numerical instabilities that make it hard to get reliable results? I've been looking into different equations of state to use for associating chemicals, and I am trying to decide if I should ditch PC-SAFT as a good option when I am calculating derivative properties for this very reason.

**Comment by ianhbell** [comment]
Actually... I have implemented the association contributions for the SAFT models. See some docs here: https://teqp.readthedocs.io/en/latest/models/Association.html .  Also implemented that for association + SAFT-VR-Mie. I added a generic SAFT container type, though the docs are, currently, not written.

The challenge with internal iterations is that teqp (and others) work with numerical types that are mostly a real numerical value along with some sort of helper type that keeps track of derivatives. That logic applies for whether you use complex step derivatives or forward-mode derivatives. Then we need to try to answer a riddle like is this true or not to determine whether we have converged:  1e-10+1e-100j < 1e-8 ? There is not an unambiguous answer to that for complex numbers and the same problem happens with autodiff types. You can get a reasonable answer in this context by making the convergence condition something like : real(1e-10+1e-100j) < 1e-8 . That is why the getbaseval() function exists in teqp. Not sure if that makes things clearer or less clear :)
